sdks/go/pkg/beam/runners/dataflow/dataflowlib/translate.go

```diff
@@ -0,0 +1,347 @@
+		return []*df.Step{x.newStep(id, windowIntoKind, prop)}, nil
+
+	case pubsub_v1.PubSubPayloadURN:
+		// Translate to native handling of PubSub I/O.
+
+		var msg pubsub_v1.PubSubPayload
+		if err := proto.Unmarshal(t.Spec.Payload, &msg); err != nil {
+			return nil, fmt.Errorf("bad pubsub payload: %v", err)
+		}
+
+		prop.Format = "pubsub"
+		prop.PubSubTopic = msg.GetTopic()
+		prop.PubSubSubscription = msg.GetSubscription()
+		prop.PubSubIDLabel = msg.GetIdAttribute()
+		prop.PubSubTimestampLabel = msg.GetTimestampAttribute()
+		prop.PubSubWithAttributes = msg.GetWithAttributes()
+
+		if prop.PubSubSubscription != "" {
+			prop.PubSubTopic = ""
+		}
+
+		switch msg.Op {
+		case pubsub_v1.PubSubPayload_READ:
+			return []*df.Step{x.newStep(id, readKind, prop)}, nil
+
+		case pubsub_v1.PubSubPayload_WRITE:
+			in := stringx.SingleValue(t.Inputs)
+
+			prop.ParallelInput = x.pcollections[in]
+			prop.Encoding = x.wrapCoder(x.comp.Pcollections[in], coder.NewBytes())
+			return []*df.Step{x.newStep(id, writeKind, prop)}, nil
+
+		default:
+			return nil, fmt.Errorf("bad pubsub op: %v", msg.Op)
+		}
+
+	default:
+		// TODO: graphx.URNCombinePerKey:
+
+		if len(t.Subtransforms) > 0 {
+			return x.translateTransforms(fmt.Sprintf("%v%v/", trunk, path.Base(t.UniqueName)), t.Subtransforms)
+		}
+
+		return nil, fmt.Errorf("unexpected primitive urn: %v", t)
+	}
+}
+
+func (x *translator) newStep(id, kind string, prop properties) *df.Step {
+	step := &df.Step{
+		Name:       id,
+		Kind:       kind,
+		Properties: newMsg(prop),
+	}
+	if prop.PubSubWithAttributes {
+		// Hack to add a empty-value property for PubSub IO. This
+		// will make PubSub send the entire message, not just
+		// the payload.
+		prop.PubSubWithAttributes = false
+		step.Properties = newMsg(propertiesWithPubSubMessage{properties: prop})
+	}
+	return step
+}
+
+func (x *translator) translateOutputs(outputs map[string]string) []output {
+	var ret []output
+	for _, out := range outputs {
+		pcol := x.comp.Pcollections[out]
+		ref := x.pcollections[out]
+
+		info := output{
+			UserName:   ref.OutputName,
+			OutputName: ref.OutputName,
+			Encoding:   x.translateCoder(pcol, pcol.CoderId),
+		}
+		ret = append(ret, info)
+	}
+	if len(ret) == 0 {
+		// Dataflow seems to require at least one output. We insert
+		// a bogus one (named "bogus") and remove it in the harness.
+
+		ret = []output{{
+			UserName:   "bogus",
+			OutputName: "bogus",
+			Encoding:   x.bogusCoderRef,
+		}}
+	}
+	return ret
+}
+
+func (x *translator) translateCoder(pcol *pb.PCollection, id string) *graphx.CoderRef {
+	c, err := x.coders.Coder(id)
+	if err != nil {
+		panic(err)
+	}
+	return x.wrapCoder(pcol, c)
+}
+
+func (x *translator) wrapCoder(pcol *pb.PCollection, c *coder.Coder) *graphx.CoderRef {
+	// TODO(herohde) 3/16/2018: ensure windowed values for Dataflow
+
+	ws := x.comp.WindowingStrategies[pcol.WindowingStrategyId]
+	wc, err := x.coders.WindowCoder(ws.WindowCoderId)
+	if err != nil {
+		panic(err)
```

Comment:
Consider printing the various IDs or data we have available here to better track down what caused the error, since it won't be obvious from the stack trace itself.
The error itself might not have the context (eg. the WindowCoder function will probably printo out the passed in ID, but it wouldn't know which PCollection and coder triggered it it.  Same for the other panic below.

Reply:
Done